zeppelin-zengine/src/main/java/org/apache/zeppelin/notebook/Notebook.java

```diff
@@ -109,6 +119,60 @@ public void recoveryIfNecessary() {
     }
   }
 
+  /**
+   * Subsystems can add a consumer, which is executed during initialization.
+   * Initialization is performed in parallel and with multiple threads.
+   *
+   * The consumer must be thread safe.
+   *
+   * @param initConsumer Consumer, which is passed the NoteId.
+   */
+  public void addInitConsumer(Consumer<String> initConsumer) {
+    this.initConsumers.add(initConsumer);
+  }
+
+  /**
+   * Asynchronous and parallel initialization of notes (during startup)
+   */
+  public void initNotebook() {
+    initExecutor = new ThreadPoolExecutor(0, Runtime.getRuntime().availableProcessors(), 1, TimeUnit.MINUTES,
```

Comment:
nit. Why don't you initialize it on the constructor or a static variable? If we call `initNotebook` multiple times by mistake, it can leak memory because `ThreaPoolExecutor` won't be shut down clearly.

Reply:
The main reason is that in the method `waitForFinishInit` the ExecutorService is shut down to wait for the termination afterwards. When calling `initNotebook` again, a new ExecutorService must be created, because the old one might be closed.
I will add a termination check in `initNotebook` to solve this problem.

Reply:
Understood the intention. Thank you for the explanation.